### PR TITLE
drivers: Clean up NXP MCUX LPSPI driver

### PR DESCRIPTION
**Pull request by decsny**
Driver was totally unreadable and tangled up in all sorts of conditional compilations and duplicated code, this PR addresses it, and makes the paths way more obvious to facilitate making changes much easier

No major functional changes intended by this PR.

There are a lot of commits so that one merge conflict doesnt cause a giant problem when rebasing.

### CODE REVIEW

**Comment by decsny** [comment]
dropped the dma path cleanup commit for now because something after recent merged patch makes the dma transfers wrong by one byte, not sure why yet

**Review by EmilioCBen — COMMENTED** [review]
Thank you for your effort on this PR,
the only concern I have is of those checks
I mentioned otherwise LGTM!

File: drivers/spi/spi_mcux_lpspi.c [hunk 1]
```diff
@@ -603,7 +608,7 @@ static int spi_mcux_transceive(const struct device *dev, const struct spi_config
 #ifdef CONFIG_SPI_MCUX_LPSPI_DMA
 	const struct spi_mcux_data *data = dev->data;
 
-	if (data->dma_rx.dma_dev && data->dma_tx.dma_dev) {
+	if (lpspi_inst_has_dma(data)) {
```
[the remark this answers is not in the document]
Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c:655]: sorry, I don't understand what did you mean, which commit ?

File: drivers/spi/spi_mcux_lpspi.c [hunk 2]
```diff
@@ -272,7 +272,10 @@ static int spi_mcux_configure(const struct device *dev, const struct spi_config
 }
 
 #ifdef CONFIG_SPI_MCUX_LPSPI_DMA
-static int spi_mcux_dma_rxtx_load(const struct device *dev, size_t *dma_size);
+static bool lpspi_inst_has_dma(const struct spi_mcux_data *data)
+{
+	return !!(data->dma_tx.dma_dev && data->dma_rx.dma_dev);
+}
```
[the remark this answers is not in the document]
Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c:239]: probably not,  I can remove that

File: drivers/spi/spi_mcux_lpspi.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c:655]: this is just using the helper function created for this check, since this check happens in multiple places, it's easier to read the code with this helper function

[the remark this answers is not in the document]
Comment (EmilioCBen) [reply drivers/spi/spi_mcux_lpspi.c:655]: Here you are checking 
if(data != data)
and the original code checked 
if(rx.dma_dev && tx.dma_dev)

[the remark this answers is not in the document]
Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c:655]: im not sure where you are seeing that, what happened here is that I abstracted that expression that previously was the condition into a helper function

[the remark this answers is not in the document]
Comment (EmilioCBen) [reply drivers/spi/spi_mcux_lpspi.c:655]: Never mind this comment, I was looking at line 547

**Review by EmilioCBen — DISMISSED** [review]

File: drivers/spi/spi_mcux_lpspi.c [hunk 3]
```diff
@@ -739,29 +744,42 @@ static void spi_mcux_iodev_complete(const struct device *dev, int status)
...
+		LOG_ERR("%s device is not ready", dma_dev->name);
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int lpspi_dma_devs_ready(struct spi_mcux_data *data)
+{
+	return lpspi_dma_dev_ready(data->dma_tx.dma_dev) |
```
Comment (danieldegrasse) [inline drivers/spi/spi_mcux_lpspi.c:716]: This is doing a binary OR on the return value of both functions. Right now this seems like it will work fine, because `-ENODEV | -ENODEV == -ENODEV`. However if the `lpspi_dma_dev_ready` function added multiple negative return codes for some reason, this would stop working. It seems like we should just return a bool from `lpspi_dma_dev_ready` and return `-ENODEV` in this function, rather than doing something tricky here.

**Review by danieldegrasse — DISMISSED** [review]
Few non-blocking nits

File: drivers/spi/spi_mcux_lpspi.c [hunk 4]
```diff
@@ -540,7 +543,9 @@ static int transceive_dma(const struct device *dev, const struct spi_config *spi
 
 	return ret;
 }
-#endif
+#else
+#define lpspi_inst_has_dma(arg) arg != arg
```
[the remark this answers is not in the document]
Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c:508]: because the arg has to be used otherwise warning

File: drivers/spi/spi_mcux_lpspi.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c:239]: it's how the spi_context code does these types of checks with returns, is why I did it

File: drivers/spi/spi_mcux_lpspi.c (hunk 3 above)
Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c:716]: both of these functions are helper functions in this file, the function is defined 2 lines above this if that happens, the person making such a change will handle it, I'm not worried about this

Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c:716]: also, the signature of device_is_ready returns a bool, and im not sure what else that lpspi_dma_dev_ready would do besides check if the device is ready, so this seems like an extremely far fetched scenario

File: drivers/spi/spi_mcux_lpspi.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (danieldegrasse) [reply drivers/spi/spi_mcux_lpspi.c:508]: Ahh, that does make sense. Thanks

**Review by dleach02 — DISMISSED** [review]

**Comment by decsny** [comment]
found the bug in the DMA cleanup commit, re--added

**Comment by decsny** [comment]
rebased and added change to fix RT platforms from upstream

File: drivers/spi/spi_mcux_lpspi.c [hunk 5]
```diff
@@ -244,44 +244,46 @@ static void spi_mcux_dma_callback(const struct device *dev, void *arg, uint32_t
...
-		if (data->ctx.tx_len == 0 && data->ctx.rx_len == 0) {
-			spi_context_complete(&data->ctx, spi_dev, 0);
-		}
-		return;
 	}
 #endif
+
+	goto done;
+error:
+	LOG_ERR("DMA callback error with channel %d.", channel);
```
Comment (danieldegrasse) [inline drivers/spi/spi_mcux_lpspi.c:282]: We haven't done this in the past (so maybe something to leave for another PR, since this PR makes no functional changes)- but it seems like we should be calling `spi_context_complete` with a nonzero status when an error occurs, right?

**Review by danieldegrasse — DISMISSED** [review]

**Comment by decsny** [comment]
@tbursztyka

File: drivers/spi/spi_mcux_lpspi.c
Comment (yvanderv) [inline drivers/spi/spi_mcux_lpspi.c:1]: c6a1daedbf91aea031a58bf9882ab508ef863cc5: please fix the typo in your git commit message - 'definitin'

**Review by yvanderv — COMMENTED** [review]
Consider comments on the git commit messages of this PR

File: drivers/spi/spi_mcux_lpspi.c
Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c:1]: > https://github.com/zephyrproject-rtos/zephyr/commit/4fa061e6ff0bcd89edbb3388a20322f54c18b844: suggest to change your summary line to better reflect what the commit is about

i only get 50 characters to work with here , I changed it but i dont know what else to say, i used the maximum number of characters in the new revision

> https://github.com/zephyrproject-rtos/zephyr/commit/41ed57c60a8c8889620be4743804b425613f969e: make the summary line a complete sentence. Bottom of what?

there are not enough characters left to say "bottom of file" , so i put "end" of file, even though the commit message says already "bottom of file" , so im not sure why someone would be confused, now it uses the maximum number of characters

> https://github.com/zephyrproject-rtos/zephyr/commit/fd1daae6326d5ae4b0308051fb928c4298df52d4: please clarify one liner, wrap?

there are not enough characters left to type the full word "wrappers", so i changed it to "calls", which is the maximum number of characters

> https://github.com/zephyrproject-rtos/zephyr/commit/b3de6553518bfde70ccb01732c6c2e0a148988b9: summary line is not clear, make it a sentence.

there are not enough characters for a sentence, so i changed it to "ErrLog dev in iodev_start" which is using the maximum number of characters

> https://github.com/zephyrproject-rtos/zephyr/commit/04f52cac07a9bff4ca3457e3dd24ad4a24ebacff: please add func or function to the summary line.

I cannot do that because the commit title already is using the maximum number of characters allowed, it should be obvious that prepare_start is a function name, especially if one reads the commit message that says "remove unneeded prepare_start function" in the first line

**Review by danieldegrasse — APPROVED** [review]

File: drivers/spi/spi_mcux_lpspi.c [hunk 6]
```diff
@@ -550,7 +553,9 @@ static int transceive_dma(const struct device *dev, const struct spi_config *spi
 
 	return ret;
 }
-#endif
+#else
+#define lpspi_inst_has_dma(arg) arg != arg
```
Comment (mmahadevan108) [inline drivers/spi/spi_mcux_lpspi.c]: Is the `#else`  required? This function is only needed when `CONFIG_SPI_MCUX_LPSPI_DMA` is defined.

**Comment by mmahadevan108** [comment]
@decsny, how was this PR tested?

File: drivers/spi/spi_mcux_lpspi.c (hunk 6 above)
Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c]: yes otherwise there will be a build error

**Comment by decsny** [comment]
> @decsny, how was this PR tested?

i ran all combinations of rtio, async, dma configs on the spi loopback test on mcxn to see if it had the same behavior as before

**Review by mmahadevan108 — APPROVED** [review]

**Comment by mmahadevan108** [comment]
@tbursztyka , can you help review
